File: gems/canvas_i18nliner/js/generate_js.js
/*
 * Copyright (C) 2014 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

var I18nliner = require("i18nliner").default;
var TranslateCall = require("i18nliner/dist/lib/extractors/translate_call").default;
var Commands = I18nliner.Commands;
var Check = Commands.Check;
var mkdirp = require("mkdirp");
var fs = require("fs");

/*
 * GenerateJs determines what needs to go into each i18n js bundle (one
 * per "i18n!scope"), based on the I18n.t calls in the code
 *
 * outputs a json file containing a mapping of scopes <-> translation keys,
 * e.g.
 *
 * {
 *   "users": [
 *     "users.title",
 *     "users.labels.foo",
 *     "foo_bar_baz" // could be from a different scope, if called within the users scope
 *   ],
 *   "groups:" [
 *     ...
 *   ],
 *   ...
 *
 */

function GenerateJs(options) {
  Check.call(this, options)
}

GenerateJs.prototype = Object.create(Check.prototype);
GenerateJs.prototype.constructor = GenerateJs;

GenerateJs.prototype.run = function() {
  var translationsWas = TranslateCall.prototype.translations
  TranslateCall.prototype.translations = function() {
    var key = this.key;
    var defaultValue = this.defaultValue;
  
    if (typeof defaultValue === 'string' || !defaultValue)
      return [[key, defaultValue]];
  
    var translations = [];
    for (var k in defaultValue) {
      if (defaultValue.hasOwnProperty(k)) {
        translations.push([key + "." + k, defaultValue[k]]);
      }
    }
    return translations;
  };
  var success = Check.prototype.run.call(this);
  if (!success) return false;
  var keysByScope = this.translations.keysByScope();
  this.outputFile = './' + (this.options.outputFile || "config/locales/generated/js_bundles.json");
  mkdirp.sync(this.outputFile.replace(/\/[^\/]+$/, ''));
  fs.writeFileSync(this.outputFile, JSON.stringify(keysByScope));
  TranslateCall.prototype.translations = translationsWas
  return true;
};

module.exports = GenerateJs;

File: gems/i18n_tasks/lib/i18n_tasks/generate_js.rb
# frozen_string_literal: true

#
# Copyright (C) 2022 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

module I18nTasks
  class GenerateJs
    PLURAL_KEYWORDS = %w[zero one two few many other].freeze

    attr_reader :index

    def initialize(index:)
      @index = index
    end

    # Compile a map of all the translations with fully-qualified keys that
    # include the scope when it's relevant (i.e. non-inferred keys.)
    #
    # Translations MUST be loaded for this to work correctly!
    #
    # Phrases are sorted by their key, and ones that do not have a translation
    # are omitted.
    def translations(locale)
      pool = core_translations(locale)

      index.each do |phrase|
        if (translation = lookup(locale, phrase["key"]))
          pool[phrase["key"]] = translation
        end
      end

      pool = pluralize(pool)
      pool.sort.to_h
    end

    private

    def pluralize(pool)
      result = Hash.new { |hash, key| hash[key] = {} }

      pool.each do |key, value|
        key_components = key.split(".")
        plural_keyword = key_components.last
        if PLURAL_KEYWORDS.include? plural_keyword
          new_key = key_components[0...-1].join(".")
          result[new_key][plural_keyword] = value
        else
          result[key] = value
        end
      end

      result
    end

    def core_translations(locale)
      pool = {}

      CORE_KEYS.map(&:to_s).each do |scope|
        phrases = lookup(locale, scope) || {}
        phrases.flatten_keys.each do |key, value|
          pool["#{scope}.#{key}"] = value
        end
      end

      pool
    end

    def lookup(locale, key)
      ::I18n.translate!(key.to_sym, locale:, raise: true)
    rescue ::I18n::MissingTranslationData
      nil
    end
  end
end
